### ISSUE TITLE
TreeView: `SelectedNodeChanged`-Event is fired when there should be no change

### ISSUE THREAD

**Issue by WolfgangKluge** [Megabit/Blazorise#4413]
**Describe the bug**
I have a TreeView and whenever the user want's to change the current node, I want to ask the user before taking any action.

**To Reproduce**
Here is a simple example

```razor
<TreeView TNode="string" Nodes="nodes" SelectedNode="selectedNode" SelectedNodeChanged="SelectedNodeChanged">
    <NodeContent>@context</NodeContent>
</TreeView>
Selected: @selectedNode
@code {
    [Inject] IMessageService MessageService { get; set; } = default!;

    string?[] nodes = new[] { "A", "B", "C" };
    string? selectedNode = "A";

    private async Task SelectedNodeChanged(string? newSelectedNode)
    {
        // is fired twice (so fast you'll never see the first message)
        if (!await MessageService.Confirm($"Navigate from {selectedNode} to {newSelectedNode}?")) return;
        selectedNode = newSelectedNode;
    }

    private async Task SelectedNodeChangedSwitchBack(string? newSelectedNode)
    {
        var prevNode = selectedNode;
        selectedNode = newSelectedNode;
        if (!await MessageService.Confirm($"Navigate from {prevNode} to {newSelectedNode}?")) {
            selectedNode = prevNode; // fires the event again (since it differs from the `state.SelectedNode`)
        }
    }
}
```

Whenever you click on any other node than the selected, a message pops up.

I expect the message shown states something like `Navigate from A to B?` and if you debug that, there is really such a message - but right after the first message, there's a second one stating `Navigate from A to A?`. At the end nothing happens.

If you use `SelectedNodeChangedSwitchBack`, the behavior is somewhat different and it seems to work - but then you get two messages in a row.

I think, the main reason for the whole problem is, that you invoke `SelectedNodeChanged` whenever `SelectedNode` is changed - even if that change comes from outside.
https://github.com/Megabit/Blazorise/blob/019c5803cf201f452172fe47329e64727c6472f0/Source/Extensions/Blazorise.TreeView/TreeView.razor.cs#L77-L92, especially the line `SelectedNodeChanged.InvokeAsync( state.SelectedNode );`

If I move that logic (except `DirtyClasses`) to `SelectNode` I get the expected behavior.

**Expected behavior**
I expect to change `selectedNode` without getting `SelectedNodeChanged` fired.

You should only fire events, if you are changing the values (not when the value is changed from outside your component). `InvokeAsync` inside the property setter of `[Parameters]` seems no good idea (for all components, not only the TreeView).
https://github.com/dotnet/aspnetcore/issues/26230

**Additional context**
My workaround is to test, if the `newSelectedNode` is changed at all - but this seems unexpected and unnecessary.

**Comment by WolfgangKluge** [comment]
Maybe the same like #2380 (but it depends on the implementation of `SelectedNodeChanged`)

**Comment by David-Moreira** [comment]
Thanks for additionally noting this:
![image](https://user-images.githubusercontent.com/22283161/209199449-01f76c30-b78c-4491-bf9d-43e7e8b7126b.png)
We've actually just touched upon this recently, we are thinking of changing and centralizing what needs changing on SetParametersAsync with the actual async support, and according to the provided Blazor guidelines. I'll go ahead and open a refactoring issue right now.

**Comment by David-Moreira** [comment]
@stsrki did you fix this on your Tree View Changes PR?

**Comment by stsrki** [comment]
I have tested @WolfgangKluge in the new version, and the new TreeView changes work. That is, it shows only one dialog. I assume that is what is needed?

One thing that still doesn't work is when the user clicks on `Cancel` button, the last selected node still stays with the selected background color. I will try to fix that before release, otherwise, it will have to wait for a patch.

**Comment by stsrki** [comment]
@WolfgangKluge any chance you could clone the repository and test the fixes in the `rel-1.2-treeview-selectednode-fixes` branch?
